pkg/cert/cert_authority.go

```diff
@@ -0,0 +1,73 @@
+package cert
+
+import (
+	"fmt"
+	"github.com/golang/glog"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+)
+
+type CertificateAuthority struct {
+	Certificate []byte
+	PrivateKey  []byte
+}
+
+// Load takes a given path to what is presumed to be a valid certificate authority and attempts to load both the
+// certificate and private key. If the path is a directory then it is assumed that there are two files "ca.crt" and
+// "ca.key" which represent the certificate and private key respectively.
+//
+// If the argument is not a directory, it must be a file ending in either of the .crt or .key extensions. Load will
+// read the given file and attempt to load the other associated file. For example, if the path given is /path/to/my-ca.crt,
+// then load will attempt to load the private key file at /path/to/my-ca.key.
+func Load(caPath string) (*CertificateAuthority, error) {
+	glog.Infof("Loading certificate authority from %v", caPath)
+	certPath, keyPath, err := certPathToCertAndKeyPaths(caPath)
+	if err != nil {
+		return nil, err
+	}
+	certMaterial, err := ioutil.ReadFile(certPath)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read cert %v: %v", certPath, err)
+	}
+	keyMaterial, err := ioutil.ReadFile(keyPath)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read key %v: %v", keyPath, err)
+	}
+	ca := CertificateAuthority{
+		Certificate: certMaterial,
+		PrivateKey:  keyMaterial,
+	}
+	return &ca, nil
+}
+
+func certPathToCertAndKeyPaths(caPath string) (string, string, error) {
+	fi, err := os.Stat(caPath)
+	if err != nil {
+		return "", "", err
+	}
+	var certPath, keyPath string
+	if fi.IsDir() {
+		certPath = filepath.Join(caPath, "ca.crt")
+		keyPath = filepath.Join(caPath, "ca.key")
+	} else {
+		ext := filepath.Ext(caPath)
+		switch ext {
+		case ".crt":
```

Comment:
This seems to imply you could give the path to just the key or just the cert. However, the code at 66 and 69 will panic if both are not specified. Is this logic specifying only one path to a file functional?

Reply:
Yes, the logic for only specifying one is functional. Note there are tests in cert_authority_test that validate this, see the following:
1. TestCertPath - validates that if the path is to the certificate file, the Load function correctly finds the private key file.
2. TestKeyPath - validates if the path is to the key file, the Load function correctly finds the certificate file.

Also there are tests that validate that proper error messages are returned if only one of the two files is present.

Reply:
I see what you are doing now. Makes sense.

Reply:
Would a lot of this be simplified by just requiring the public and private key to be in a single PEM file?

Reply:
It would make it simpler for us, but I would caution that it would make using this feature more difficult for the user. Right now they have three options: the path to the directory in which the cert and PK live, the path to the cert, or the path to the PK. 

In addition, I think we *should* support separate cert & PK files, because that is the default way in which they are generated by kubeadm, so for someone copying the CA off an existing cluster or trying to reproduce that cluster, they would have them in seperate files.